File: JsonSchemaAndStructs/KiteConnect-Python-DataClass/mf_sips.py
# This code parses date/times, so please
#
#     pip install python-dateutil
#
# To use this code, make sure you
#
#     import json
#
# and then, to convert JSON from a string, do
#
#     result = mf_sips_from_dict(json.loads(json_string))

from dataclasses import dataclass
from typing import Optional, Dict, Any, List, TypeVar, Callable, Type, cast
from datetime import datetime
import dateutil.parser


T = TypeVar("T")


def from_none(x: Any) -> Any:
    assert x is None
    return x


def from_str(x: Any) -> str:
    assert isinstance(x, str)
    return x


def from_union(fs, x):
    for f in fs:
        try:
            return f(x)
        except:
            pass
    assert False


def from_int(x: Any) -> int:
    assert isinstance(x, int) and not isinstance(x, bool)
    return x


def from_datetime(x: Any) -> datetime:
    return dateutil.parser.parse(x)


def from_dict(f: Callable[[Any], T], x: Any) -> Dict[str, T]:
    assert isinstance(x, dict)
    return { k: f(v) for (k, v) in x.items() }


def is_type(t: Type[T], x: Any) -> T:
    assert isinstance(x, t)
    return x


def from_list(f: Callable[[Any], T], x: Any) -> List[T]:
    assert isinstance(x, list)
    return [f(y) for y in x]


def to_class(c: Type[T], x: Any) -> dict:
    assert isinstance(x, c)
    return cast(Any, x).to_dict()


@dataclass(slots=True)
class Datum:
    sip_reg_num: Optional[int] = None
    completed_instalments: Optional[int] = None
    created: Optional[datetime] = None
    dividend_type: Optional[str] = None
    frequency: Optional[str] = None
    fund: Optional[str] = None
    instalment_amount: Optional[int] = None
    instalment_day: Optional[int] = None
    instalments: Optional[int] = None
    last_instalment: Optional[datetime] = None
    next_instalment: Optional[datetime] = None
    pending_instalments: Optional[int] = None
    sip_id: Optional[str] = None
    sip_type: Optional[str] = None
    status: Optional[str] = None
    step_up: Optional[Dict[str, int]] = None
    tag: Optional[str] = None
    tradingsymbol: Optional[str] = None
    transaction_type: Optional[str] = None
    trigger_price: Optional[int] = None

    @staticmethod
    def from_dict(obj: Any) -> 'Datum':
        assert isinstance(obj, dict)
        sip_reg_num = from_union([from_none, lambda x: int(from_str(x))], obj.get("sip_reg_num"))
        completed_instalments = from_union([from_int, from_none], obj.get("completed_instalments"))
        created = from_union([from_datetime, from_none], obj.get("created"))
        dividend_type = from_union([from_str, from_none], obj.get("dividend_type"))
        frequency = from_union([from_str, from_none], obj.get("frequency"))
        fund = from_union([from_str, from_none], obj.get("fund"))
        instalment_amount = from_union([from_int, from_none], obj.get("instalment_amount"))
        instalment_day = from_union([from_int, from_none], obj.get("instalment_day"))
        instalments = from_union([from_int, from_none], obj.get("instalments"))
        last_instalment = from_union([from_datetime, from_none], obj.get("last_instalment"))
        next_instalment = from_union([from_datetime, from_none], obj.get("next_instalment"))
        pending_instalments = from_union([from_int, from_none], obj.get("pending_instalments"))
        sip_id = from_union([from_str, from_none], obj.get("sip_id"))
        sip_type = from_union([from_str, from_none], obj.get("sip_type"))
        status = from_union([from_str, from_none], obj.get("status"))
        step_up = from_union([lambda x: from_dict(from_int, x), from_none], obj.get("step_up"))
        tag = from_union([from_str, from_none], obj.get("tag"))
        tradingsymbol = from_union([from_str, from_none], obj.get("tradingsymbol"))
        transaction_type = from_union([from_str, from_none], obj.get("transaction_type"))
        trigger_price = from_union([from_int, from_none], obj.get("trigger_price"))
        return Datum(sip_reg_num, completed_instalments, created, dividend_type, frequency, fund, instalment_amount, instalment_day, instalments, last_instalment, next_instalment, pending_instalments, sip_id, sip_type, status, step_up, tag, tradingsymbol, transaction_type, trigger_price)

    def to_dict(self) -> dict:
        result: dict = {}
        result["sip_reg_num"] = from_union([lambda x: from_none((lambda x: is_type(type(None), x))(x)), lambda x: from_str((lambda x: str((lambda x: is_type(int, x))(x)))(x))], self.sip_reg_num)
        result["completed_instalments"] = from_union([from_int, from_none], self.completed_instalments)
        result["created"] = from_union([lambda x: x.isoformat(), from_none], self.created)
        result["dividend_type"] = from_union([from_str, from_none], self.dividend_type)
        result["frequency"] = from_union([from_str, from_none], self.frequency)
        result["fund"] = from_union([from_str, from_none], self.fund)
        result["instalment_amount"] = from_union([from_int, from_none], self.instalment_amount)
        result["instalment_day"] = from_union([from_int, from_none], self.instalment_day)
        result["instalments"] = from_union([from_int, from_none], self.instalments)
        result["last_instalment"] = from_union([lambda x: x.isoformat(), from_none], self.last_instalment)
        result["next_instalment"] = from_union([lambda x: x.isoformat(), from_none], self.next_instalment)
        result["pending_instalments"] = from_union([from_int, from_none], self.pending_instalments)
        result["sip_id"] = from_union([from_str, from_none], self.sip_id)
        result["sip_type"] = from_union([from_str, from_none], self.sip_type)
        result["status"] = from_union([from_str, from_none], self.status)
        result["step_up"] = from_union([lambda x: from_dict(from_int, x), from_none], self.step_up)
        result["tag"] = from_union([from_str, from_none], self.tag)
        result["tradingsymbol"] = from_union([from_str, from_none], self.tradingsymbol)
        result["transaction_type"] = from_union([from_str, from_none], self.transaction_type)
        result["trigger_price"] = from_union([from_int, from_none], self.trigger_price)
        return result


@dataclass(slots=True)
class MFSips:
    data: Optional[List[Datum]] = None

    @staticmethod
    def from_dict(obj: Any) -> 'MFSips':
        assert isinstance(obj, dict)
        data = from_union([lambda x: from_list(Datum.from_dict, x), from_none], obj.get("data"))
        return MFSips(data)

    def to_dict(self) -> dict:
        result: dict = {}
        result["data"] = from_union([lambda x: from_list(lambda x: to_class(Datum, x), x), from_none], self.data)
        return result


def mf_sips_from_dict(s: Any) -> MFSips:
    return MFSips.from_dict(s)


def mf_sips_to_dict(x: MFSips) -> Any:
    return to_class(MFSips, x)

File: JsonSchemaAndStructs/KiteConnect-Typescript-Structs/mf_sips.ts
// To parse this data:
//
//   import { Convert, MFSips } from "./file";
//
//   const mFSips = Convert.toMFSips(json);
//
// These functions will throw an error if the JSON doesn't
// match the expected interface, even if the JSON is valid.

export interface MFSips {
    data?: Datum[];
}

export interface Datum {
    completedInstalments?: number;
    created?:              Date;
    dividendType?:         string;
    frequency?:            string;
    fund?:                 string;
    instalmentAmount?:     number;
    instalmentDay?:        number;
    instalments?:          number;
    lastInstalment?:       Date;
    nextInstalment?:       Date;
    pendingInstalments?:   number;
    sipID?:                string;
    sipRegNum?:            null | string;
    sipType?:              string;
    status?:               string;
    stepUp?:               { [key: string]: number };
    tag?:                  string;
    tradingsymbol?:        string;
    transactionType?:      string;
    triggerPrice?:         number;
}

// Converts JSON strings to/from your types
// and asserts the results of JSON.parse at runtime
export class Convert {
    public static toMFSips(json: string): MFSips {
        return cast(JSON.parse(json), r("MFSips"));
    }

    public static mFSipsToJson(value: MFSips): string {
        return JSON.stringify(uncast(value, r("MFSips")), null, 2);
    }
}

function invalidValue(typ: any, val: any, key: any = ''): never {
    if (key) {
        throw Error(`Invalid value for key "${key}". Expected type ${JSON.stringify(typ)} but got ${JSON.stringify(val)}`);
    }
    throw Error(`Invalid value ${JSON.stringify(val)} for type ${JSON.stringify(typ)}`, );
}

function jsonToJSProps(typ: any): any {
    if (typ.jsonToJS === undefined) {
        const map: any = {};
        typ.props.forEach((p: any) => map[p.json] = { key: p.js, typ: p.typ });
        typ.jsonToJS = map;
    }
    return typ.jsonToJS;
}

function jsToJSONProps(typ: any): any {
    if (typ.jsToJSON === undefined) {
        const map: any = {};
        typ.props.forEach((p: any) => map[p.js] = { key: p.json, typ: p.typ });
        typ.jsToJSON = map;
    }
    return typ.jsToJSON;
}

function transform(val: any, typ: any, getProps: any, key: any = ''): any {
    function transformPrimitive(typ: string, val: any): any {
        if (typeof typ === typeof val) return val;
        return invalidValue(typ, val, key);
    }

    function transformUnion(typs: any[], val: any): any {
        // val must validate against one typ in typs
        const l = typs.length;
        for (let i = 0; i < l; i++) {
            const typ = typs[i];
            try {
                return transform(val, typ, getProps);
            } catch (_) {}
        }
        return invalidValue(typs, val);
    }

    function transformEnum(cases: string[], val: any): any {
        if (cases.indexOf(val) !== -1) return val;
        return invalidValue(cases, val);
    }

    function transformArray(typ: any, val: any): any {
        // val must be an array with no invalid elements
        if (!Array.isArray(val)) return invalidValue("array", val);
        return val.map(el => transform(el, typ, getProps));
    }

    function transformDate(val: any): any {
        if (val === null) {
            return null;
        }
        const d = new Date(val);
        if (isNaN(d.valueOf())) {
            return invalidValue("Date", val);
        }
        return d;
    }

    function transformObject(props: { [k: string]: any }, additional: any, val: any): any {
        if (val === null || typeof val !== "object" || Array.isArray(val)) {
            return invalidValue("object", val);
        }
        const result: any = {};
        Object.getOwnPropertyNames(props).forEach(key => {
            const prop = props[key];
            const v = Object.prototype.hasOwnProperty.call(val, key) ? val[key] : undefined;
            result[prop.key] = transform(v, prop.typ, getProps, prop.key);
        });
        Object.getOwnPropertyNames(val).forEach(key => {
            if (!Object.prototype.hasOwnProperty.call(props, key)) {
                result[key] = transform(val[key], additional, getProps, key);
            }
        });
        return result;
    }

    if (typ === "any") return val;
    if (typ === null) {
        if (val === null) return val;
        return invalidValue(typ, val);
    }
    if (typ === false) return invalidValue(typ, val);
    while (typeof typ === "object" && typ.ref !== undefined) {
        typ = typeMap[typ.ref];
    }
    if (Array.isArray(typ)) return transformEnum(typ, val);
    if (typeof typ === "object") {
        return typ.hasOwnProperty("unionMembers") ? transformUnion(typ.unionMembers, val)
            : typ.hasOwnProperty("arrayItems")    ? transformArray(typ.arrayItems, val)
            : typ.hasOwnProperty("props")         ? transformObject(getProps(typ), typ.additional, val)
            : invalidValue(typ, val);
    }
    // Numbers can be parsed by Date but shouldn't be.
    if (typ === Date && typeof val !== "number") return transformDate(val);
    return transformPrimitive(typ, val);
}

function cast<T>(val: any, typ: any): T {
    return transform(val, typ, jsonToJSProps);
}

function uncast<T>(val: T, typ: any): any {
    return transform(val, typ, jsToJSONProps);
}

function a(typ: any) {
    return { arrayItems: typ };
}

function u(...typs: any[]) {
    return { unionMembers: typs };
}

function o(props: any[], additional: any) {
    return { props, additional };
}

function m(additional: any) {
    return { props: [], additional };
}

function r(name: string) {
    return { ref: name };
}

const typeMap: any = {
    "MFSips": o([
        { json: "data", js: "data", typ: u(undefined, a(r("Datum"))) },
    ], false),
    "Datum": o([
        { json: "completed_instalments", js: "completedInstalments", typ: u(undefined, 0) },
        { json: "created", js: "created", typ: u(undefined, Date) },
        { json: "dividend_type", js: "dividendType", typ: u(undefined, "") },
        { json: "frequency", js: "frequency", typ: u(undefined, "") },
        { json: "fund", js: "fund", typ: u(undefined, "") },
        { json: "instalment_amount", js: "instalmentAmount", typ: u(undefined, 0) },
        { json: "instalment_day", js: "instalmentDay", typ: u(undefined, 0) },
        { json: "instalments", js: "instalments", typ: u(undefined, 0) },
        { json: "last_instalment", js: "lastInstalment", typ: u(undefined, Date) },
        { json: "next_instalment", js: "nextInstalment", typ: u(undefined, Date) },
        { json: "pending_instalments", js: "pendingInstalments", typ: u(undefined, 0) },
        { json: "sip_id", js: "sipID", typ: u(undefined, "") },
        { json: "sip_reg_num", js: "sipRegNum", typ: u(undefined, u(null, "")) },
        { json: "sip_type", js: "sipType", typ: u(undefined, "") },
        { json: "status", js: "status", typ: u(undefined, "") },
        { json: "step_up", js: "stepUp", typ: u(undefined, m(0)) },
        { json: "tag", js: "tag", typ: u(undefined, "") },
        { json: "tradingsymbol", js: "tradingsymbol", typ: u(undefined, "") },
        { json: "transaction_type", js: "transactionType", typ: u(undefined, "") },
        { json: "trigger_price", js: "triggerPrice", typ: u(undefined, 0) },
    ], false),
};
